### ISSUE TITLE
False positive for `B031` in mutually exclusive branches

### ISSUE THREAD

**Issue by janosh** [charliermarsh/ruff#3801]
Ruff v0.0.260

```py
# t.py
import itertools

for key, group in itertools.groupby("foobar"):
    if True:
        print(key, len(tuple(group)))
    else:
        print(key, len(tuple(group)))  # B031
```

```sh
ruff t.py --select B
```

raises false positive `flake8-bugbear` B031

> Using the generator returned from `itertools.groupby()` more than once will do nothing on the second usage

**Comment by charliermarsh** [comment]
Is this incorrect? `flake8-bugbear` flags line 7 too.

**Comment by janosh** [comment]
If the generator uses are in mutually exclusive execution paths, I think this error should not be raised.

**Comment by charliermarsh** [comment]
Ahh yeah, I see. It might be difficult to detect this reliably.

**Comment by charliermarsh** [comment]
(But yes, it's a false positive.)
